### PR TITLE
[Caffe2] Update hip files

### PR DESCRIPTION
**Pull request by rohithkrn**
The goal of this PR is to update the hip files to reflect relevant changes in cuda source files.

### CODE REVIEW

**Comment by rohithkrn** [comment]
@bddppq  This also adds THCCachingAllocator for hip. But, `THCCachingAllocator_hip.cc` includes `THCCachingAllocator.h` has code specific to cuda. Could we rename that to `THCCachingAllocator_gpu.h`, so that we can add corresponding hip file.

@petrex

File: caffe2/core/hip/context_hip.h [hunk 1]
```diff
@@ -51,7 +52,7 @@ class ThreadLocalHIPObjects {
 
   hipStream_t GetStream(int gpu, int stream_id) {
     vector<hipStream_t>& gpu_streams = hip_streams_[gpu];
-    if (gpu_streams.size() <= stream_id) {
+    if (gpu_streams.size() <= (unsigned)stream_id) {
```
Comment (petrex) [inline caffe2/core/hip/context_hip.h]: Can we use static_cast<> here?

File: caffe2/operators/hip/softmax_op_miopen.cc [hunk 2]
```diff
@@ -44,6 +44,10 @@ class MIOpenSoftmaxOp final : public Operator<HIPContext> {
     const int D = X.size_from_dim(canonical_axis);
 
     Y->ResizeLike(X);
+    auto* Y_data = Y->template mutable_data<T>();
```
Comment (petrex) [inline caffe2/operators/hip/softmax_op_miopen.cc]: I am assuming no real functional change in *miopen. Let me know if that is not the case thx.

**Comment by petrex** [comment]
@bddppq  is there an update on the base docker img?

`Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/caffe2/conda3-cuda9.0-cudnn7-ubuntu16.04:126 not found`

**Review by petrex — CHANGES_REQUESTED** [review]

**Comment by ezyang** [comment]
There is currently an outage for caffe2 rocm builds. It will be fixed in the next hour.

File: caffe2/operators/hip/softmax_op_miopen.cc (hunk 2 above)
Comment (rohithkrn) [reply caffe2/operators/hip/softmax_op_miopen.cc]: It adds a check for empty batch size, i.e N=0, before calling any miopen functions.

File: caffe2/core/hip/context_hip.h (hunk 1 above)
Comment (rohithkrn) [reply caffe2/core/hip/context_hip.h]: We can, I was just imitating the cuda file.

File: caffe2/operators/hip/softmax_op_miopen.cc (hunk 2 above)
Comment (rohithkrn) [reply caffe2/operators/hip/softmax_op_miopen.cc]: Updated the test script.

File: caffe2/core/hip/THCCachingAllocator_hip.cc [hunk 3]
```diff
@@ -0,0 +1,309 @@
+#include "caffe2/core/THCCachingAllocator.h"
```
[the remark this answers is not in the document]
Comment (rohithkrn) [reply caffe2/core/hip/THCCachingAllocator_hip.cc]: It has a conflict with one of the mappings, nonetheless we would prefer to not hipify files under core.

**Comment by bddppq** [comment]
better to hipify THCCachingAllocator files

**Comment by bddppq** [comment]
@rohithkrn I just checked there are only two places in aten using cudaError instead of cudaErorr_t, I think we can simply change those two places and remove the "cudaError" -> "hipError_t" mapping

**Comment by rohithkrn** [comment]
@bddppq I have changed the mapping order to make it work in the current state. But yes, changing aten also should work.

**Review by bddppq — APPROVED** [review]
LG

**Pull request closed without merging**